rfcs/react-components/convergence/theme-shared-colors.md

```diff
@@ -0,0 +1,114 @@
+
+**The numbers confirm the performance decreases linearly depending on the number of CSS variables (and therefore theme tokens).**
+
+### New Shared Color Alias tokens request
+
+_The numbers might be inaccurate as the feature is not fully spec-ed yet!_
+
+In order to support new scenarios (colorful calendar view), there is a new request to add 3 new shared color alias tokens. Adding them would result in adding 3 alias tokens for each of 49 shared colors = adding 147 new shared color alias tokens which would affect application performance.
+
+## Detailed Design or Proposal
+
+The theme currently contains 49 global shared colors.
+
+Current alias structure forces each ramp to be uniform (the same set of alias tokens for each shared color), resulting in 49 alias tokens each time we need a new line item (new alias token). Today we have 9 alias slots = total alias count is 441 (49x9).
+
+Components have varying degrees of color needs.
+
+Instead of 1 uniform Shared alias table (today's experience), separate out into 3 tables:
+
+1. Semantic States
+   - There are 5 unique global colors needed to support this set (Danger, Success, Severe, Warning, Out of office)
+   - Assume 9 alias tokens (Could potentially remove 2 from this table as they were avatar specific)
+   - 45 alias tokens in total (5 x 9)
+2. Persona
+   - There are 30 unique global colors needed to support this set
+   - Only needs 3 alias tokens
+   - 90 tokens in total (30 x 3)
+3. Category (calendar categories) (not fully defined yet)
+   - 10 unique global colors to support this set
+   - Assuming 9 alias slots (subject to increase)
+   - 90 tokens in total (10 x 9)
+
+Dividing up by 3 tables would reduce token count from 441 to ~225 (Table 1+Table 2+Table 3) while adding support for Calendar scenario.
+
+### Pros and Cons
+
+- 👍 Reducing shared color alias tokens from 441 to ~225 instead of increasing them to ~600 (to add tokens needed for Calendar)
+- 👍 More clarity around which specific color to use for semantic states instead of relying on design guidance
+- 👍 Allows a bit more wiggle room for the semantic alias table to grow w/o adding too much bloat
+- 👎 **Breaking change after the final RC**
+- 👎 Three separate color sets instead of a single one
+
+## Discarded Solutions
+
+### Create `theme_v2` to maintain backwards compatibility
+
+Instead of breaking the current theme we can just deprecate it and add `_v2` version.
+
+That might cause partner confusion - releasing the first final version with deprecated things which, when used, would result in degraded performance. The breaking change should not affect partners much - the colors which we are removing are not supposed to be used anywhere. If those are, there will be a build time (TS) error to catch the problem quickly.
+
+### Other possible optimization
+
+There is still a longer term plan to explore partial themes, theme splitting and lazy loading. Those do not conflict with reducing the number of tokens now.
+
```

Comment:
```suggestion
#### Option: ColorSetProvider
1. Take the color palettes (16 color ramps) from global and define each in an exported module and const objects.
2. Remove color palettes (9 color ramps) from theme
3. Add a ColorSetProvider component that can define a set of CSS vars with a pattern.
This component would take a palette array of colors and a name.
It would set cssVars for use in colorful scenarios:
- color`name`Background1
- color`name`Background2
- color`name`Background3
- color`name`BorderActive
- color`name`Border1
- color`name`Border2
- color`name`Foreground1
- color`name`Foreground2
- color`name`Foreground3
4. Update Avatar and other components that want to support color components to replace the colorful boolean property with a colorSetName.  This would cause them to reference the colors set by a ColorSetProvider upstream.

Pros and Cons
(Y) Color palettes in own modules should tree shake out if not used.
(Y) Default case does not add any of the color set CSS vars, keeping the theme CSS vars minimal.
(Y) ColorSetProvider can map to position in the palette based on theme if necessary -or- we can leave updating the color palette array at the top level when the theme changes.
(Y) With reduces impact on the system, could extend the color sets to include hover, active, and selected states.  Could also go to 1 background, border, and foreground color.  Note: Should rename border to stroke!
(N) There is some fragility for consumers that have extensions that reference color sets and have to rely on their host to have provided them.  
(N) Unsure of the setting of color sets at scopes lower in the hierarchy.  Option: Could integrate colorSets as a property into FluentProvider to encourage setting at higher levels.
```